### PR TITLE
#2308 Select a neighborhood consider parent path

### LINKED ISSUE #2308: Selecting a neighbor in the neighborhood view: Consider "parent path"
Consider this use case:

image

The second neighbor of the selected neighborhood points to `Organism|ArterialBlood_UmbilicalCord|BloodCells`. While `Organism`is not present in this module, it is set as the parent of `ArterialBlood_UmbilicalCord`, so the full path could be resolved and the container shown in the right pane.

### PR DESCRIPTION
**Pull request by benjaperez1983**
Fixes #2308

# Description

 Select a neighborhood consider parent path

## Type of change

Please mark relevant options with an `x` in the brackets.

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires documentation changes (link at least one [user](https://github.com/Open-Systems-Pharmacology/docs) or [developer](https://github.com/Open-Systems-Pharmacology/developer-docs) documentation issue):
- [ ] Algorithm update - updates algorithm documentation/questions/answers etc.
- [ ] Other (please describe):

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Integration tests
- [x] Unit tests
- [ ] Manual tests 
- [ ] No tests required

# Reviewer checklist

Mark everything that needs to be checked before merging the PR.

- [ ] Check if the code is well documented
- [ ] Check if the behavior is what is expected
- [ ] Check if the code is well tested
- [ ] Check if the code is readable and well formatted
- [ ] Additional checks (document below if any)
- [ ] Check if documentation update issue(s) are created if the option `This change requires a documentation update` above is selected

# Screenshots (if appropriate):

# Questions (if appropriate):

### CODE REVIEW

File: src/OSPSuite.Core/Domain/ObjectPath.cs [hunk 1]
```diff
@@ -146,6 +149,15 @@ public virtual T Resolve<T>(IEntity refEntity) where T : class
          return resolvePath<T>(dependentObject, usePath);
       }
 
+      private void removeMatchingPaths(IContainer root, List<string> usePath)
+      {
+         var matchingElements = !string.IsNullOrEmpty(root?.ParentPath?.PathAsString) &&
+                                new ObjectPath(_pathEntries).PathAsString.StartsWith(root?.ParentPath?.PathAsString ?? string.Empty);
```
Comment (Yuri05) [inline src/OSPSuite.Core/Domain/ObjectPath.cs:155]: the part after `&&` could be simplified to 
`new ObjectPath(_pathEntries).PathAsString.StartsWith(root.ParentPath.PathAsString);`
(because the part before `&&` already ensures that `root.ParentPath.PathAsString` is not null and not empty)

**Review by rwmcintosh — APPROVED** [review]
I would just rename that one method a little more explicitly.

File: tests/OSPSuite.Core.Tests/Domain/ObjectPathSpecs.cs [hunk 2]
```diff
@@ -116,6 +117,39 @@ public void should_return_the_desired_entity()
...
+         }
+      }
+
+      [Observation]
+      public void should_return_the_desired_entity()
+      {
+         sut.Resolve<IParameter>(_organ).ShouldBeEqualTo(_parameter);
+      }
+
+      private static string GenerateRandomString(Random random, int length)
```
Comment (msevestre) [inline tests/OSPSuite.Core.Tests/Domain/ObjectPathSpecs.cs:145]: I have no idea why we need to generate random strings suddenly in a text. I have no idea what's going on.
Please create a path that we can analyze visually and not in a weird loop.

File: src/OSPSuite.Core/Domain/ObjectPath.cs [hunk 3]
```diff
@@ -146,6 +149,15 @@ public virtual T Resolve<T>(IEntity refEntity) where T : class
          return resolvePath<T>(dependentObject, usePath);
       }
 
+      private void removeMatchingParentPath(IContainer root, List<string> usePath)
+      {
+         var matchingElements = !string.IsNullOrEmpty(root?.ParentPath?.PathAsString) &&
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/ObjectPath.cs:154]: same `root?.ParentPath?.PathAsString` locally in a variable  and check that it's not null or empty then return

File: src/OSPSuite.Core/Domain/ObjectPath.cs [hunk 4]
```diff
@@ -146,6 +149,15 @@ public virtual T Resolve<T>(IEntity refEntity) where T : class
...
+                                new ObjectPath(_pathEntries).PathAsString.StartsWith(root?.ParentPath?.PathAsString ?? string.Empty);
```
Comment (msevestre) [inline src/OSPSuite.Core/Domain/ObjectPath.cs:155]: that ?? can never happen since you check before

Comment (msevestre) [reply src/OSPSuite.Core/Domain/ObjectPath.cs:155]: the code is kind of weird. we are using _pathEntries (global) to modify the parameter pass (usePath). 
this works because usePath is created using _pathEntries but this is really hard to follow. 
Also we have lists, we got to string, we compare strings and we then remove from list.
Should't we just use...no string at all?

**Review by Yuri05 — APPROVED** [review]
